extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java

```diff
@@ -369,12 +373,40 @@ interface RestClientConfig {
          */
         Optional<String> url();
 
+        /**
+         * Duplicate mapping of {@link RestClientConfig#url()} to keep a reference of the name used to retrieve the
+         * <code>url</code>. We need this to reload the <code>url</code> configuration in case it contains an expression
+         * to <code>${quarkus.http.port}</code>, which is only set after we load the config.
+         */
+        @ConfigDocIgnore
+        @WithName("url")
+        ConfigValue urlValue();
```

Comment:
Maybe it's a crazy idea but should we affect the random port when reading the config instead? Maybe by using a converter?

Reply:
Unfortunately, no. The problem here is how we read and set `quarkus.http.port`.

One of the first things we do when we start Quarkus is to load the configurations and cache the values in the mappings / roots (this is the same for old roots and new mappings). This is fine when we explicitly set the `quarkus.http.port` since the value gets expanded to the explicitly set port.

When we set a random port of `0`, we delegate the random port assignment to Vert.x, but we also need the configuration to be loaded to start Vert.x. So we have a chicken-egg problem here because the configuration starts before Vert.x, any expansion done for `${quarkus.http.port}` is `0` (the random port). This will get cached in the config objects.

Once Vert.x sets the port, this is mutating the `SystemProperties`:
https://github.com/quarkusio/quarkus/blob/f399671722319495d93349ecd9ed1337ddc218ea/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L1312-L1332

Which is a different problem that is causing other issues like:
https://github.com/quarkusio/quarkus/issues/42844
https://github.com/quarkusio/quarkus/pull/42106#issuecomment-2250045346
https://github.com/quarkusio/quarkus/issues/27996

When we get the actual port, all the configuration objects are cached, and we need to query the configuration again to get the updated value. The problem is getting the discovered property name by the REST Client (see explanation in https://github.com/quarkusio/quarkus/pull/42904#issuecomment-2323280321). Adding a `ConfigValue` here at the mapping class allows us to keep a reference to the full property name of the `URL' so we can query it without having to reconstruct all the REST Client name variations.

This worked previously because it was specifically implemented in REST Client with:
https://github.com/quarkusio/quarkus/pull/21530

Reply:
But is Vert.x doing things that we can't do ourselves when the config is read? The contract of this is that you get a random port, we don't really care if it's Vert.x assigning it? Especially since you might want a random port for something else than Vert.x.

Now maybe Vert.x is doing something very clever that we can't replicate easily at the config level?

Reply:
BTW, if it wasn't obvious, I'm not advocating that we change things now. But I'm wondering if it's worth discussing it for the future.

Reply:
Also obviously, we would need to keep a list of the ports randomly affected around to make sure we don't affect the same port twice.

Reply:
I fully agree. We need to remove the random port generation from Vert.x and do it before the config is set up. This would eliminate many of our current hacks and issues.

Of course, I didn't want to make that change here, so I had to rely on another hack :)